### PR TITLE
Fix for Issue #569 (CROSS JOIN)  

### PR DESCRIPTION
**Pull request by ili**
instead of #575 

fix for issues #569, #588

### CODE REVIEW

File: Source/Linq/Builder/SelectManyBuilder.cs [hunk 1]
```diff
@@ -129,11 +131,42 @@ protected override IBuildContext BuildMethodCall(ExpressionBuilder builder, Meth
...
+				var subquery  = false;
+				var tables    = sequence.SelectQuery.From.Tables;
+				var baseTable = tables[0];
+
+				// if new join has dependency to many From tables we have to convert them to INNER JOINS
+				if (tables.Count > 1)
+				{
+					if (builder.DataContextInfo.SqlProviderFlags.IsCrossJoinSupported || builder.DataContextInfo.SqlProviderFlags.IsInnerJoinAsCrossSupported)
+					{
+						for (var i = tables.Count - 1; i > 0; i--)
```
Comment (sdanyliv) [inline Source/Linq/Builder/SelectManyBuilder.cs]: Bad idea to change order.

Comment (ili) [inline Source/Linq/Builder/SelectManyBuilder.cs]: Why? Order should not matter, or I'm mistaken?

**Comment by ili** [comment]
Yes, #569 & #588 are the same

File: Source/Linq/Builder/SelectManyBuilder.cs (hunk 1 above)
Comment (sdanyliv) [inline Source/Linq/Builder/SelectManyBuilder.cs]: Better to keep, it is more easier in future to analyze linq expression and its SQL.
You can rewrite it by clearing collection after processing and add single basetable.

File: Source/Linq/Builder/SelectManyBuilder.cs [hunk 2]
```diff
@@ -129,11 +131,45 @@ protected override IBuildContext BuildMethodCall(ExpressionBuilder builder, Meth
...
-				var join = leftJoin ? SelectQuery.OuterApply(sql) : SelectQuery.CrossApply(sql);
-				sequence.SelectQuery.From.Tables[0].Joins.Add(join.JoinedTable);
-
-				context.Collection = new SubQueryContext(collection, sequence.SelectQuery, false);
-				return new SelectContext(buildInfo.Parent, resultSelector, sequence, context);
+				var join = leftJoin ? SelectQuery.OuterApply(sql) : SelectQuery.CrossApply(sql);
				sequence.SelectQuery.From.Tables[0].Joins.Add(join.JoinedTable);

				context.Collection = new SubQueryContext(collection, sequence.SelectQuery, false);
				return new SelectContext(buildInfo.Parent, resultSelector, sequence, context);
```
Comment (sdanyliv) [inline Source/Linq/Builder/SelectManyBuilder.cs]: I think, you have to just revert this part of code.

File: Source/SqlProvider/BasicSqlOptimizer.cs [hunk 3]
```diff
@@ -37,12 +37,143 @@ public virtual SelectQuery Finalize(SelectQuery selectQuery)
...
+
+					var sc =                  new QueryVisitor().Convert(query.Select,  changeSource) ?? query.Select;
+					var ic = query.IsInsert ? new QueryVisitor().Convert(query.Insert,  changeSource) ?? query.Insert : null;
+					var uc = query.IsUpdate ? new QueryVisitor().Convert(query.Update,  changeSource) ?? query.Update : null;
+					var dc = query.IsDelete ? new QueryVisitor().Convert(query.Delete,  changeSource) ?? query.Delete : null;
+					var wc =                  new QueryVisitor().Convert(query.Where,   changeSource) ?? query.Where;
+					var gc =                  new QueryVisitor().Convert(query.GroupBy, changeSource) ?? query.GroupBy;
+					var hc =                  new QueryVisitor().Convert(query.Having,  changeSource) ?? query.Having;
+					var oc =                  new QueryVisitor().Convert(query.OrderBy, changeSource) ?? query.OrderBy;
+					var us = query.HasUnion ? query.Unions.Select(_ => new QueryVisitor().Convert(_, changeSource)).ToList() : query.Unions;
```
Comment (sdanyliv) [inline Source/SqlProvider/BasicSqlOptimizer.cs]: Thi all can be done by convertnig full query but not it's parts.

File: Source/SqlProvider/BasicSqlOptimizer.cs [hunk 4]
```diff
@@ -37,12 +37,143 @@ public virtual SelectQuery Finalize(SelectQuery selectQuery)
...
+
+					var ps = new List<SqlParameter>(query.Parameters.Count);
+
+					ps.AddRange(query.Parameters);
+
+					query.Init(ic, uc, dc, sc, query.From, wc, gc, hc, oc, us,
```
Comment (sdanyliv) [inline Source/SqlProvider/BasicSqlOptimizer.cs]: If you convert full query, this part is not needed.

File: Source/SqlProvider/BasicSqlOptimizer.cs [hunk 5]
```diff
@@ -37,12 +37,143 @@ public virtual SelectQuery Finalize(SelectQuery selectQuery)
...
 				selectQuery = OptimizeJoins(selectQuery);
 
 			return selectQuery;
 		}
 
+		SelectQuery BuildCrossJoins(SelectQuery selectQuery)
+		{
+			bool setAlias = false;
+
+			new QueryVisitor().Visit(selectQuery, element =>
```
Comment (sdanyliv) [inline Source/SqlProvider/BasicSqlOptimizer.cs]: Instead of Visit, use Convert here.

**Pull request merged**